magento/pwa-studio
# PWA-718: FormError

## Description

As a shopper, I would like to see helpful and consistent messaging when I encounter an error in the app. This will allow me to know how to recover from those errors if possible, and ideally lead to fewer abandoned carts and support requests sent to the merchant.

In scope of PWA-336, we identified areas in CartPage and CheckoutPage that were lacking error handling, and consolidated the various way we were handling errors in the app. In reviewing this with UX, Product, and Team Lead, we identified some consistent patterns we would like to start enforcing in the app. In review of this task, we also discovered some bugs in how we were parsing errors from GraphQL that will also need to be resolved (should happen organically with addressing the following).

- ~~Implement FormError throughout the app~~
- ~~Update UX for FormError to new mocks provided by Scott Hensley~~
- ~~Implement utility function/hook that will handle parsing GraphQL Errors~~
- ~~Be sure all components rendering errors not using FormError use this utility (should really only be Toasts)~~
- Given that GraphQL errors are not very user friendly, we need to come up with an ErrorMapper to render better errors (out of scope?)
- ~~Consistently handle/don't handle re-throwing these errors. We have a lot of places where we catch and console.error, which doesn't really accomplish anything, since Apollo has its own error boundary.~~
- ~~When writing tests for FormError, it proved difficult to test the scrollIntoView effect due to jsdom limitations. Stephen Rugh found this article that outlines a reasonable alternative to consider that makes this testable, and giving us another custom hook to add. Worth discussing with team, but its an option so we can have coverage on that features outside of E2E testing.~~ Added in 362e57c0878119202bf5eaabc3462baff768c7fd
 
## Related Issue





Closes [PWA-718](https://jira.corp.magento.com/browse/PWA-718).

## Acceptance 



### Verification Stakeholders

### Specification



### Verification Steps


1. For each form in the app, verify that there is consistent error handling.
2. Verify nothing is broken :D

## Screenshots / Screen Captures (if appropriate)

**Form Error w/o left border, below heading of form**

[![Image from Gyazo](https://i.gyazo.com/a3e4c6d6280cac1f66410df53a37b138.png)](https://gyazo.com/a3e4c6d6280cac1f66410df53a37b138)

## Checklist

* I have updated the documentation accordingly, if necessary.
